### DISCUSSION TITLE
PromQL: additions and subtractions between some vectors and an empty vector result in empty vector

### DISCUSSION THREAD

**Discussion by dgrisonnet** [prometheus/prometheus#10207] [Q&A] [answered]
**What did you do?**

I encountered a corner case in a PromQL expression used in an alert where some vector - an empty vector resulted in an empty vector. This may be intended, but it seemed odd as it would make it impossible to do additions and subtractions on optional vectors.

**What did you expect to see?**

I expected additions and subtractions between a vector and an empty vector to return the original vector.

**What did you see instead? Under which circumstances?**

Instead, I saw that an empty vector was returned.

The exact scenario where I noticed this oddity was for computing the number of Kubernetes kube-apiserver requests that are above a 500ms latency threshold. The query was:

```
sum (apiserver_request_duration_seconds_count{job="apiserver",verb=~"LIST|GET", scope="namespace"})
-
sum (apiserver_request_duration_seconds_bucket{job="apiserver",verb=~"LIST|GET", scope="namespace", le="0.5"})
```

It was working perfectly fine until the second part of the expression started returning an empty vector after updating to a new version of Kubernetes. This was because the `0.5` second was removed and so the second part of the query wasn't matching any timeseries anymore.

This is easily reproducible with additions and subtractions with the following queries:
```
sum(http_requests_total)
-
metric_without_timeseries
```
```
sum(http_requests_total)
+
metric_without_timeseries
```

It seemed pretty harmless to me at first sight, but after thinking about it a bit more, this behavior is quite problematic. For example, when we want to compute the number of requests that are too slow for SLO-based alerting, we may want to define different latency thresholds depending on the endpoint that is hit and the verb of the request. To achieve that we would have the following promQL expression:

```
sum(rate(http_requests_duration_seconds_count[1h]))
-
(
  sum(rate(http_requests_duration_seconds_bucket{path="/a", le="0.1"}[1h]))
  +
  sum(rate(http_requests_duration_seconds_bucket{path="/b", verb=~"GET|LIST", le="0.1"}[1h]))
  +
  sum(rate(http_requests_duration_seconds_bucket{path="/b", verb=~"POST|DELETE", le="0.2"}[1h]))
  +
  sum(rate(http_requests_duration_seconds_bucket{path!~"/{ab}", le="0.5"}
)
```

If any of these expressions return an empty vector then the whole expression would result in an empty vector. This is very problematic since I would never expect any of these expressions to always return values since in the end, these metrics are dependent on clients' inputs.

I don't know how Prometheus is treating the empty vector in this particular case, but maybe what I would expect would be to have it converted to either vector(0) or 0 for substractions and additions.

**Environment**

* Prometheus version: v2.32.1

**Comment by beorn7** [comment, marked as answer]
I don't think it's a bug. It's fundamental semantics of vector matching in PromQL. Changing it would be a huge change, it would be breaking, and it would arguably do more harm than good.

In your case, the behavior might appear annoying, but I guess that's mostly because you are `sum`'ing everything up so you end up with no labels and you actually don't really want label matching. 

You can "protect" all the expressions that potentially evaluate to an empty vector with `or vector(0)`.

**Reply by dgrisonnet** [reply]
Thank you for clarifying :)

**Comment by roidelapluie** [comment]
After reading your issue twice, this is the expected behaviour. You can read more about our vector matching here: https://prometheus.io/docs/prometheus/latest/querying/operators/#vector-matching
